pageserver/src/storage_sync/index.rs

```diff
@@ -222,62 +222,142 @@ pub struct RemoteTimeline {
+            .extend(new_layers.into_iter().map(|p| (p, Default::default())));
     }
 
     pub fn add_upload_failures(&mut self, upload_failures: impl IntoIterator<Item = PathBuf>) {
-        self.missing_layers.extend(upload_failures.into_iter());
+        // FIXME
+        self.missing_layers
+            .extend(upload_failures.into_iter().map(|p| (p, Default::default())));
     }
 
     pub fn remove_layers(&mut self, layers_to_remove: &HashSet<PathBuf>) {
         self.timeline_layers
-            .retain(|layer| !layers_to_remove.contains(layer));
+            .retain(|layer, _| !layers_to_remove.contains(layer));
         self.missing_layers
-            .retain(|layer| !layers_to_remove.contains(layer));
+            .retain(|layer, _| !layers_to_remove.contains(layer));
     }
 
     /// Lists all layer files in the given remote timeline. Omits the metadata file.
-    pub fn stored_files(&self) -> &HashSet<PathBuf> {
-        &self.timeline_layers
+    pub fn stored_files(&self) -> impl Iterator<Item = &PathBuf> {
+        self.timeline_layers.keys()
     }
 
     pub fn from_index_part(timeline_path: &Path, index_part: IndexPart) -> anyhow::Result<Self> {
         let metadata = TimelineMetadata::from_bytes(&index_part.metadata_bytes)?;
+        let default_metadata = &IndexLayerMetadata::default();
         Ok(Self {
-            timeline_layers: to_local_paths(timeline_path, index_part.timeline_layers),
-            missing_layers: to_local_paths(timeline_path, index_part.missing_layers),
+            timeline_layers: index_part
+                .timeline_layers
+                .iter()
+                .map(|path| {
+                    (
+                        path.as_path(timeline_path),
+                        index_part
+                            .layer_metadata
+                            .get(path)
+                            .unwrap_or(default_metadata)
+                            .into(),
+                    )
+                })
+                .collect(),
+            missing_layers: index_part
+                .missing_layers
+                .iter()
+                .map(|path| {
+                    (
+                        path.as_path(timeline_path),
+                        index_part
+                            .layer_metadata
+                            .get(path)
+                            .unwrap_or(default_metadata)
+                            .into(),
+                    )
+                })
+                .collect(),
             metadata,
             awaits_download: false,
         })
     }
 }
 
+/// Metadata gathered for each of the layer files.
+///
+/// Fields have to be `Option`s because remote [`IndexPart`]'s can be from different version, which
+/// might have less or more metadata depending if upgrading or rolling back an upgrade.
+#[derive(Debug, Clone, Default, PartialEq, Eq, PartialOrd, Ord)]
+pub struct LayerFileMetadata {
+    file_size: Option<u64>,
```

Comment:
Builder doesn't seem to be really needed so far. I feel that wherever these values need to be created is usually close to layer file download or creation, which has access to all of the metadata we can envision. It's just a matter of minor reshuffling to fit everything in (thinking of file hashing).

Stealing this thread to discuss Option<T> inside `LayerFileMetadata`. In general we'd want to avoid having them there, because it introduces unwanted matching, but the issue is I cannot see a way around it.

Specifically related to this I can see the problem of `/v1/tenant/:tenant_id/attach` which scans remote storage and needs to end up with a `RemoteIndex` even with an older `IndexPart`.

Unsure how to work around this.

- Trying to solve this with `IndexPart::Legacy` and `IndexPart::V1` doesn't help, because it matters for `RemoteIndex`.
- `HashMap<PathBuf, Option<LayerFileMetadata>>` within `RemoteIndex` doesn't really work with incremental updates to `LayerFileMetadata`, like the easily envisioned next step of SHA-2 256 file hash
    - One would need versioned `LayerFileMetadata`, but the API still needs to have `Option<u64>` and `Option<FileHash>` (tbd)
    - Handling this option at the boundary to the `remote_storage` [around `pageserver::storage_sync::upload`](https://github.com/neondatabase/neon/pull/2582/files#diff-d1c451c1f0ea40c9743827ce26f9aa6e8f51f1591530c7793596b9a9dc9599d4R134-R143) will still have problematic Option::None

Reply:
See https://github.com/neondatabase/neon/pull/2582#issuecomment-1277472252. Resolving for now.